components/external_storage/export/src/service.rs

```diff
@@ -0,0 +1,68 @@
+use crate::export::create_storage;
+use external_storage::{ExternalStorage};
+use std::io::{ErrorKind};
+
+use grpcio::{self, *};
+use kvproto::backup as proto;
+use std::collections::HashMap;
+
+type Storages = HashMap<String, Box<dyn ExternalStorage>>;
+
+/// Service handles the RPC messages for the `ExternalStorage` service.
+pub struct Service {
+    storages: Storages,
+}
+
+impl Service {
+    /// Create a new backup service.
+    pub fn new(storages: Storages) -> Service {
+        Service { storages }
+    }
+}
+
+impl proto::ExternalStorage for Service {
+    fn start(
+        &mut self,
+        _ctx: RpcContext,
+        req: proto::ExternalStorageStartRequest,
+        sink: UnarySink<proto::ExternalStorageStartResponse>,
+    ) {
+        match create_storage(req.get_storage_backend()) {
+            Ok(storage) => {
+                let name = storage.name().to_string();
+                // This allows just one aws storage
+                // But we could allow multiple
+                // This insert will update any existing storage
+                self.storages.insert(name.clone(), storage);
+                let mut rsp = proto::ExternalStorageStartResponse::default();
+                rsp.set_storage_id(name);
+                sink.success(rsp);
+            },
+            Err(e) => {
+                sink.fail(make_rpc_error(e));
+            },
+        }
+    }
+
+    fn finish(
+        &mut self,
+        _ctx: RpcContext,
+        req: proto::ExternalStorageFinishRequest,
+        sink: UnarySink<proto::ExternalStorageFinishResponse>,
+    ) {
+        let storage_id = req.get_storage_id();
+        self.storages.remove(storage_id);
+        sink.success(proto::ExternalStorageFinishResponse::default());
+    }
```

Comment:
What other methods look such as writing data to a file? Maybe a multipart uploading interface should be exposed

Reply:
yes, currently this is incomplete and I am looking into the upload/download part.

Reply:
The experimental GRPC service is completed now.